### ISSUE TITLE
Install provenance tracking

### ISSUE THREAD

**Issue by dommitchell** [DOAJ/doaj#1044]
As we have opened up the site to the ambassadors, we need to become cleverer about how we monitor who is doing what.

We would like to install provenance tracking. If it is feasible to do this, we would need to have it installed fairly soon.

Can I have your thoughts on the feasibility and timing please?

**Comment by dommitchell** [comment]
Linking this to #544 as I think there are related, right?

**Comment by dommitchell** [comment]
Dom to create list of what we want to report on.

**Comment by dommitchell** [comment]
***DRAFT, not complete***

As requested, some examples of reports that we would like to see:

- number of records (apps or reapps) touched by one user
- number of records fully processed to a conclusion by a user
- number of records (apps or reapps) touched by an editorial group
- number of records (apps or reapps) processed to a conclusion by an editorial group
- number of applications added per country per year

**Comment by richard-jones** [comment]
Proposing that we add a new index type, called "provenance" or similar, which contains objects of the following structure:

```json
{
    "id" : "<provenance record id>",
    "created_date" : "<when this action took place>",
    "user": "<user that carried out the action>",
    "editor_group": ["<list of editor groups the user was in at the time>"],
    "type" : "<type being acted on: suggestion, journal, etc>",
    "action" : "<string denoting the action taken on the object>",
    "resource_id" : "<id of the type being acted on>",
}
```

**Comment by dommitchell** [comment]
I notice that in Section B above I didn't add a desired common denominator, that we should be able to break everything down by country, eg number of journals added per month per country, number of article added per year per country, etc. Does that add a complicated layer?

**Comment by richard-jones** [comment]
Section (B) above is duplicated in #544 so it is fine to close this issue when the current Provenance work has been completed.

**Comment by richard-jones** [comment]
* Note that when counting "number of records" means "unique" records
* "fully processed" = rejected, accepted, completed, ready

maned can do rejected or accepted
editors can do ready
associated editors can do completed

**Comment by richard-jones** [comment]
One more quite subtle question occurs.

In the meeting we decided that what it means to count the "number of records" was the number of unique records, so we only counted 2 edits of the same application as one edit.

As we're breaking these reports down by time period, and counting, say, by month, how do we count if a user has edited the same item in different time periods?

Suppose a user edits item 12345 in May 2017 and then again in June 2017.  What is the report?  Is it:

May 2017 - 1 edit
June 2017 - 1 edit

or is it

May 2017 - 1 edit
June 2017 - 0 edits

the currently plan would see us do the latter, which is the most consistent.  If we do the former, there's a chance that sum of the numbers for all the months would not be the same as the same report broken down by year.

**Comment by greboun** [comment]
The purpose of the excercise is that we link users to their activities on
applications.
If a user  open and edits an application 5 times a month we count as 1
I would say that if a user works on an application over different months we
still count only the nmber of new applications in that month
So the second option is OK

TOM

Dr. Tom Olijhoek
Editor-in-Chief  DOAJ
Support us: www.doaj.org/support

**Comment by dommitchell** [comment]
@greboun what's the time limit that we give an AssEd or Ed to start work on a record before we take it away and reassign it to someone else?

**Comment by greboun** [comment]
@dommitchell officially it is 5 working days
How it works in practice is another thing



Dr. Tom Olijhoek
Editor-in-Chief  DOAJ
Support us: www.doaj.org/support

**Comment by dommitchell** [comment]
> the currently plan would see us do the latter, which is the most consistent.

@richard-jones most consistent with what? And so to follow the example, what happens if we extrapolate it further? What happens in these two scenarios?

May 2017 - 1 edit
June 2017 - 0 edits
July 2017 - 1 edit

May 2017 - 1 edit
June 2017 - 0 edits
July 2017 - 0 edits

Sorry if this was discussed at the meeting on Tuesday.

**Comment by greboun** [comment]
@dommitchell @richard-jones  I would say that we don't want applications
being evaluated for more than 1 month as per our policy!
So such extrapolation is somewhat irrealistic

Dr. Tom Olijhoek
Editor-in-Chief  DOAJ
Support us: www.doaj.org/support

**Comment by dommitchell** [comment]
I'd still like to know as this will be important when generating reports for Lars etc.

**Comment by richard-jones** [comment]
> the currently plan would see us do the latter, which is the most consistent.

@dommitchell - as in, most consistent with the current plan to only count multiple edits of an application as one.  If we introduced counting edits in different months as 2 edits, this would be inconsistent with counting edits within one month as 1 edit, since the granularity of the time period is essentially arbitrary.

Your scenarios raise another interesting point that I'm only just crystallising in my head.  If there are multiple edits, in which time bin do we count them?  The answer will be dependent on /when/ the report is run, and the total time period over which it runs, which makes things even more fun :)

There are two ways we could count the values for each month:

* "Count Early" - Once a record is edited for the first time, all subsequent counts are recorded for that month, even if they appeared in the month after
* "Count Late" - All edit counts are attached to the date of the most recent edit, and recorded in that month, even if they appeared in previous months.

Suppose we have 2 users who behave as per Dom's scenarios.  Each user is editing a single (different) item:

| Month | U1 Edits | U2 Edits |
| --- | --- | --- |
| May | 1 | 1 |
| June | 0 | 1 |
| July | 1 | 1 |

Now, each time we generate the report, suppose that we generate it for up to 12 months in the past, so we definitely get each month re-generated each report

In May:

| Month | U1 Number of actual edits | U1 Count earliest | U1 Count latest | U2 Number of actual edits | U2 Count earliest | U2 Count latest |
| --- | --- | --- | --- | --- | --- | --- |
| May | 1 |  1 | 1 | 1 | 1 | 1 |

In June:

| Month | U1 Number of actual edits | U1 Count earliest | U1 Count latest | U2 Number of actual edits | U2 Count earliest | U2 Count latest |
| --- | --- | --- | --- | --- | --- | --- |
| May | 1 | 1 | 1 | 1 | 2 | 0 |
| June | 0 |  0 | 0 | 1 | 0 | 2 |

In July:

| Month | U1 Number of actual edits | U1 Count earliest | U1 Count latest | U2 Number of actual edits | U2 Count earliest | U2 Count latest |
| --- | --- | --- | --- | --- | --- | --- |
| May | 1 | 2 | 0 | 1 | 3 | 0 |
| June | 0 |  0 | 0 | 1 | 0 | 0 |
| July | 1 |  0 | 2 | 1 | 0 | 3 |

So, the thing is, no matter how you approach it, you can never truly say that the count for a given month is ended, and in order to know the "right" answer you have to regenerate the report over the entire period.  You can't build the report up, month-by-month, essentially.  Which is not a big deal for us computationally, so is not a technical problem.

This might be a mountain out of a molehill, but I do need to know which way to go just to write the code one way or another.  Practically, these issues probably won't arise for associate editors, but I guess they could with editors or managing editors.  It will, of course, affect anyone working on the 31st and the 1st of each month though :)

Do I have a recommendation?  There are a whole range of options to choose from:

1. Count edits individually rather than unique edits, and the problem goes away entirely
2. Count edits binned by month, and just accept that sometimes one item will be counted in both bins
3. Count early, this helps you understand when a user started working on a record, but doesn't tell you when they finished
4. Count late, this helps you understand when a user finished working on a record, but doesn't tell you when they started.

In the case that users finish an entire record between the 1st and 31st of the month, 2, 3, and 4 are equivalent.

Options 1 and 2 don't require the months already reported on to be recalculated in the future.

Options 3 and 4 will always require an entire spread of months to be calculated, and those months must always be considered together as a whole.  Any months not generated at the same time may be inconsistent with the months generated at the time.

Overall, my feeling is now for (2), for simplicity and pragmatism.  As long as you don't want to report on small time periods like days and hours, it will probably be fine.  A bit of double counting will occur, but probably not much.  We'll also have the underlying edit provenance data to go back to if there's a question.

**Comment by richard-jones** [comment]
Note that this problem affects edits much more than it affects status changes to completion, since we imagine the former may occur regularly, while the latter is likely to happen just the once.

**Comment by emanuil-tolev** [comment]
> 2\. Count edits binned by month, and just accept that sometimes one item will be counted in both bins

:+1:

**Comment by dommitchell** [comment]
> while the latter is likely to happen just the once

Switching to completed can happen more than once but certainly happens less often than an edit. An associate editor will switch a record to completed when they think they are finished. An Editor or Managing Editor may switch it back to in progress if they think it is incomplete. I don't suppose thta makes any difference to the reporting structure?

Thanks for the detailed. I am inclined to go with option 2: a little overcounting doesn't actually matter as long as the error is consistent throughout. 

Finally, re-reading this thread I saw this comment which I added but which wasn't acknowledged and I just want to make sure it hasn't been missed: 

https://github.com/DOAJ/doaj/issues/1044#issuecomment-232068785
